🐛 BUG: Uncaught TypeError: utils.inherits is not a function

### Which Cloudflare product(s) does this pertain to?

Workers Runtime, Wrangler

### What version(s) of the tool(s) are you using?

3.80.0 [Wrangler]

### What version of Node are you using?

20.17.0

### What operating system and version are you using?

Mac Sonoma 14.6.1

### Describe the Bug

### Observed behavior

I have a worker which imports `ethers@5.7.2` library. When I switched to `compatibility_flags = ["nodejs_compat_v2"]` it started throwing:
```
✘ [ERROR] service core:user:error-example: Uncaught TypeError: utils.inherits is not a function

    at null.<anonymous> (index.js:7544:11) in
  node_modules/.pnpm/hash.js@1.1.7/node_modules/hash.js/lib/hash/sha/1.js
    at null.<anonymous> (index.js:12:50) in __require
    at null.<anonymous> (index.js:8295:20) in
  node_modules/.pnpm/hash.js@1.1.7/node_modules/hash.js/lib/hash/sha.js
    at null.<anonymous> (index.js:12:50) in __require
    at null.<anonymous> (index.js:8812:17) in
  node_modules/.pnpm/hash.js@1.1.7/node_modules/hash.js/lib/hash.js
    at null.<anonymous> (index.js:12:50) in __require
    at null.<anonymous> (index.js:15900:28)
```

### Expected behavior

The worker should start.

### Steps to reproduce

1. Clone the https://github.com/nkuba/cf-workers-error-example repository
2. Run `pnpm install`
3. Run `pnpm dev`

### Please provide a link to a minimal reproduction

https://github.com/nkuba/cf-workers-error-example

### Please provide any relevant error logs

```
> wrangler dev

 ⛅️ wrangler 3.80.0
-------------------

[wrangler:inf] Ready on http://localhost:8787
⎔ Starting local server...
✘ [ERROR] service core:user:error-example: Uncaught TypeError: utils.inherits is not a function

    at null.<anonymous> (index.js:7544:11) in
  node_modules/.pnpm/hash.js@1.1.7/node_modules/hash.js/lib/hash/sha/1.js
    at null.<anonymous> (index.js:12:50) in __require
    at null.<anonymous> (index.js:8295:20) in
  node_modules/.pnpm/hash.js@1.1.7/node_modules/hash.js/lib/hash/sha.js
    at null.<anonymous> (index.js:12:50) in __require
    at null.<anonymous> (index.js:8812:17) in
  node_modules/.pnpm/hash.js@1.1.7/node_modules/hash.js/lib/hash.js
    at null.<anonymous> (index.js:12:50) in __require
    at null.<anonymous> (index.js:15900:28)



✘ [ERROR] The Workers runtime failed to start. There is likely additional logging output above.
```

--- a/.changeset/nasty-melons-camp.md
+++ b/.changeset/nasty-melons-camp.md
@@ -1,0 +1,10 @@
+---
+"wrangler": minor
+---
+
+fix: allow `require`ing aliased npm packages
+
+Before this PR `require`ing an NPM packages aliased in unenv would fail.
+That's because `require` would load the mjs file.
+
+This PR adds wraps the mjs file in a virtual ES module to allow `require`ing it.

--- a/.changeset/nasty-melons-camp.md
+++ b/.changeset/nasty-melons-camp.md
@@ -4,7 +4,7 @@
 
 fix: allow `require`ing aliased npm packages
 
-Before this PR `require`ing an NPM packages aliased in unenv would fail.
+Before this PR `require`ing NPM packages aliased in unenv would fail.
 That's because `require` would load the mjs file.
 
 This PR adds wraps the mjs file in a virtual ES module to allow `require`ing it.

--- a/.changeset/nasty-melons-camp.md
+++ b/.changeset/nasty-melons-camp.md
@@ -2,9 +2,9 @@
 "wrangler": minor
 ---
 
-fix: allow `require`ing aliased npm packages
+fix: allow `require`ing unenv aliased packages
 
-Before this PR `require`ing NPM packages aliased in unenv would fail.
+Before this PR `require`ing packages aliased in unenv would fail.
 That's because `require` would load the mjs file.
 
 This PR adds wraps the mjs file in a virtual ES module to allow `require`ing it.

--- a/fixtures/nodejs-hybrid-app/src/index.ts
+++ b/fixtures/nodejs-hybrid-app/src/index.ts
@@ -7,6 +7,7 @@ import { Context } from "vm";
 import { Client } from "pg";
 import { s } from "./dep.cjs";
 
+
 testBasicNodejsProperties();
 
 export default {
@@ -26,17 +27,25 @@ export default {
 				return testPostgresLibrary(env, ctx);
 			case "/test-x509-certificate":
 				return testX509Certificate();
+			case "/test-require-npm":
+				return testRequireNpmAlias();
 		}
 
 		return new Response(
 			'<a href="query">Postgres query</a> | ' +
 				'<a href="test-process">Test process global</a> | ' +
 				'<a href="test-random">Test getRandomValues()</a> | ' +
-				'<a href="test-x509-certificate">Test X509Certificate</a>',
+				'<a href="test-x509-certificate">Test X509Certificate</a>' +
+				'<a href="test-require-npm">Test require aliased npm package</a>',
 			{ headers: { "Content-Type": "text/html; charset=utf-8" } }
 		);
 	},
 };
+
+function testRequireNpmAlias() {
+	const inherits = require("inherits");
+	return new Response(typeof inherits === "function" ? `"OK!"` : `"KO!"`);
+}
 
 function testX509Certificate() {
 	try {

--- a/fixtures/nodejs-hybrid-app/src/index.ts
+++ b/fixtures/nodejs-hybrid-app/src/index.ts
@@ -7,7 +7,6 @@ import { Context } from "vm";
 import { Client } from "pg";
 import { s } from "./dep.cjs";
 
-
 testBasicNodejsProperties();
 
 export default {

--- a/fixtures/nodejs-hybrid-app/src/index.ts
+++ b/fixtures/nodejs-hybrid-app/src/index.ts
@@ -26,8 +26,8 @@ export default {
 				return testPostgresLibrary(env, ctx);
 			case "/test-x509-certificate":
 				return testX509Certificate();
-			case "/test-require-npm":
-				return testRequireNpmAlias();
+			case "/test-require-alias":
+				return testRequireUenvAliasedPackages();
 		}
 
 		return new Response(
@@ -35,13 +35,13 @@ export default {
 				'<a href="test-process">Test process global</a> | ' +
 				'<a href="test-random">Test getRandomValues()</a> | ' +
 				'<a href="test-x509-certificate">Test X509Certificate</a>' +
-				'<a href="test-require-npm">Test require aliased npm package</a>',
+				'<a href="test-require-alias">Test require unenv aliased packages</a>',
 			{ headers: { "Content-Type": "text/html; charset=utf-8" } }
 		);
 	},
 };
 
-function testRequireNpmAlias() {
+function testRequireUenvAliasedPackages() {
 	const fetch = require("cross-fetch");
 	const supportsDefaultExports = typeof fetch === "function";
 	const supportsNamedExports = typeof fetch.Headers === "function";

--- a/fixtures/nodejs-hybrid-app/src/index.ts
+++ b/fixtures/nodejs-hybrid-app/src/index.ts
@@ -42,8 +42,12 @@ export default {
 };
 
 function testRequireNpmAlias() {
-	const inherits = require("inherits");
-	return new Response(typeof inherits === "function" ? `"OK!"` : `"KO!"`);
+	const fetch = require("cross-fetch");
+	const supportsDefaultExports = typeof fetch === "function";
+	const supportsNamedExports = typeof fetch.Headers === "function";
+	return new Response(
+		supportsDefaultExports && supportsNamedExports ? `"OK!"` : `"KO!"`
+	);
 }
 
 function testX509Certificate() {

--- a/fixtures/nodejs-hybrid-app/tests/index.test.ts
+++ b/fixtures/nodejs-hybrid-app/tests/index.test.ts
@@ -3,7 +3,6 @@ import { fetch } from "undici";
 import { describe, it, test } from "vitest";
 import { runWranglerDev } from "../../shared/src/run-wrangler-long-lived";
 
-
 describe("nodejs compat", () => {
 	it("should work when running code requiring polyfills", async ({
 		expect,

--- a/fixtures/nodejs-hybrid-app/tests/index.test.ts
+++ b/fixtures/nodejs-hybrid-app/tests/index.test.ts
@@ -64,13 +64,13 @@ describe("nodejs compat", () => {
 		}
 	});
 
-	test("import aliased npm packages", async ({ expect }) => {
+	test("import unenv aliased packages", async ({ expect }) => {
 		const { ip, port, stop } = await runWranglerDev(
 			resolve(__dirname, "../src"),
 			["--port=0", "--inspector-port=0"]
 		);
 		try {
-			const response = await fetch(`http://${ip}:${port}/test-require-npm`);
+			const response = await fetch(`http://${ip}:${port}/test-require-alias`);
 			await expect(response.text()).resolves.toBe(`"OK!"`);
 		} finally {
 			await stop();

--- a/fixtures/nodejs-hybrid-app/tests/index.test.ts
+++ b/fixtures/nodejs-hybrid-app/tests/index.test.ts
@@ -3,6 +3,7 @@ import { fetch } from "undici";
 import { describe, it, test } from "vitest";
 import { runWranglerDev } from "../../shared/src/run-wrangler-long-lived";
 
+
 describe("nodejs compat", () => {
 	it("should work when running code requiring polyfills", async ({
 		expect,
@@ -58,6 +59,19 @@ describe("nodejs compat", () => {
 			const response = await fetch(
 				`http://${ip}:${port}/test-x509-certificate`
 			);
+			await expect(response.text()).resolves.toBe(`"OK!"`);
+		} finally {
+			await stop();
+		}
+	});
+
+	test("import aliased npm packages", async ({ expect }) => {
+		const { ip, port, stop } = await runWranglerDev(
+			resolve(__dirname, "../src"),
+			["--port=0", "--inspector-port=0"]
+		);
+		try {
+			const response = await fetch(`http://${ip}:${port}/test-require-npm`);
 			await expect(response.text()).resolves.toBe(`"OK!"`);
 		} finally {
 			await stop();

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
@@ -6,7 +6,7 @@ import { getBasePath } from "../../paths";
 import type { Plugin, PluginBuild } from "esbuild";
 
 const REQUIRED_NODE_BUILT_IN_NAMESPACE = "node-built-in-modules";
-const REQUIRED_NPM_PACKAGE_NAMESPACE = "required-npm-package";
+const REQUIRED_UNENV_ALIAS_NAMESPACE = "required-unenv-alias";
 
 export const nodejsHybridPlugin: () => Plugin = () => {
 	const { alias, inject, external } = env(nodeless, cloudflare);
@@ -109,11 +109,12 @@ function handleUnenvAliasedPackages(
 		// Note: Does not apply to Node.js packages that are handled in `handleRequireCallsToNodeJSBuiltins`
 		if (
 			args.kind === "require-call" &&
-			(unresolvedAlias.startsWith("unenv/runtime/npm/") || unresolvedAlias.startsWith("unenv/runtime/mock/"))
+			(unresolvedAlias.startsWith("unenv/runtime/npm/") ||
+				unresolvedAlias.startsWith("unenv/runtime/mock/"))
 		) {
 			return {
 				path: args.path,
-				namespace: REQUIRED_NPM_PACKAGE_NAMESPACE,
+				namespace: REQUIRED_UNENV_ALIAS_NAMESPACE,
 			};
 		}
 		// Resolve the alias to its absolute path and potentially mark it as external
@@ -123,13 +124,30 @@ function handleUnenvAliasedPackages(
 		};
 	});
 
+	build.initialOptions.banner = { js: "", ...build.initialOptions.banner };
+	build.initialOptions.banner.js += dedent`
+		function __cf_cjs(esm) {
+		  const cjs = 'default' in esm ? esm.default : {};
+			for (const [k, v] of Object.entries(esm)) {
+				if (k !== 'default') {
+					Object.defineProperty(cjs, k, {
+						enumerable: true,
+						value: v,
+					});
+				}
+			}
+			return cjs;
+		}
+		`;
+
 	build.onLoad(
-		{ filter: /.*/, namespace: REQUIRED_NPM_PACKAGE_NAMESPACE },
+		{ filter: /.*/, namespace: REQUIRED_UNENV_ALIAS_NAMESPACE },
 		({ path }) => {
 			return {
 				contents: dedent`
-        import * as all from '${path}';
-        module.exports = 'default' in all ? all.default : all;`,
+        import * as esm from '${path}';
+				module.exports = __cf_cjs(esm);
+				`,
 				loader: "js",
 			};
 		}

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
@@ -89,21 +89,17 @@ function handleUnenvAliasedPackages(
 	external: string[]
 ) {
 	// esbuild expects alias paths to be absolute
-	const aliasAbsolute = Object.fromEntries(
-		Object.entries(alias)
-			.map(([key, value]) => {
-				let resolvedAliasPath;
-				try {
-					resolvedAliasPath = require.resolve(value);
-				} catch (e) {
-					// this is an alias for package that is not installed in the current app => ignore
-					resolvedAliasPath = "";
-				}
+	const aliasAbsolute: Record<string, string> = {};
+	for (const [module, unresolvedAlias] of Object.entries(alias)) {
+		try {
+			aliasAbsolute[module] = require
+				.resolve(unresolvedAlias)
+				.replace(/\.cjs$/, ".mjs");
+		} catch (e) {
+			// this is an alias for package that is not installed in the current app => ignore
+		}
+	}
 
-				return [key, resolvedAliasPath.replace(/\.cjs$/, ".mjs")];
-			})
-			.filter((entry) => entry[1] !== "")
-	);
 	const UNENV_ALIAS_RE = new RegExp(
 		`^(${Object.keys(aliasAbsolute).join("|")})$`
 	);

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
@@ -5,7 +5,9 @@ import { cloudflare, env, nodeless } from "unenv";
 import { getBasePath } from "../../paths";
 import type { Plugin, PluginBuild } from "esbuild";
 
+
 const REQUIRED_NODE_BUILT_IN_NAMESPACE = "node-built-in-modules";
+const REQUIRED_NPM_PACKAGE_NAMESPACE = "required-npm-package";
 
 export const nodejsHybridPlugin: () => Plugin = () => {
 	const { alias, inject, external } = env(nodeless, cloudflare);
@@ -14,7 +16,7 @@ export const nodejsHybridPlugin: () => Plugin = () => {
 		setup(build) {
 			errorOnServiceWorkerFormat(build);
 			handleRequireCallsToNodeJSBuiltins(build);
-			handleAliasedNodeJSPackages(build, alias, external);
+			handleUnenvAliasedPackages(build, alias, external);
 			handleNodeJSGlobals(build, inject);
 		},
 	};
@@ -55,7 +57,7 @@ function errorOnServiceWorkerFormat(build: PluginBuild) {
 }
 
 /**
- * We must convert `require()` calls for Node.js to a virtual ES Module that can be imported avoiding the require calls.
+ * We must convert `require()` calls for Node.js modules to a virtual ES Module that can be imported avoiding the require calls.
  * We do this by creating a special virtual ES module that re-exports the library in an onLoad handler.
  * The onLoad handler is triggered by matching the "namespace" added to the resolve.
  */
@@ -81,7 +83,7 @@ function handleRequireCallsToNodeJSBuiltins(build: PluginBuild) {
 	);
 }
 
-function handleAliasedNodeJSPackages(
+function handleUnenvAliasedPackages(
 	build: PluginBuild,
 	alias: Record<string, string>,
 	external: string[]
@@ -107,12 +109,35 @@ function handleAliasedNodeJSPackages(
 	);
 
 	build.onResolve({ filter: UNENV_ALIAS_RE }, (args) => {
+		// Convert `require()` calls for NPM packages to a virtual ES Module that can be imported avoiding the require calls.
+		// Note: Only applies to npm aliases as node package are handled in `handleRequireCallsToNodeJSBuiltins`
+		if (
+			args.kind === "require-call" &&
+			alias[args.path].startsWith("unenv/runtime/npm")
+		) {
+			return {
+				path: args.path,
+				namespace: REQUIRED_NPM_PACKAGE_NAMESPACE,
+			};
+		}
 		// Resolve the alias to its absolute path and potentially mark it as external
 		return {
 			path: aliasAbsolute[args.path],
 			external: external.includes(alias[args.path]),
 		};
 	});
+
+	build.onLoad(
+		{ filter: /.*/, namespace: REQUIRED_NPM_PACKAGE_NAMESPACE },
+		({ path }) => {
+			return {
+				contents: dedent`
+        import * as all from '${path}';
+        module.exports = 'default' in all ? all.default : all;`,
+				loader: "js",
+			};
+		}
+	);
 }
 
 /**

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
@@ -5,7 +5,6 @@ import { cloudflare, env, nodeless } from "unenv";
 import { getBasePath } from "../../paths";
 import type { Plugin, PluginBuild } from "esbuild";
 
-
 const REQUIRED_NODE_BUILT_IN_NAMESPACE = "node-built-in-modules";
 const REQUIRED_NPM_PACKAGE_NAMESPACE = "required-npm-package";
 

--- a/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
+++ b/packages/wrangler/src/deployment-bundle/esbuild-plugins/hybrid-nodejs-compat.ts
@@ -104,11 +104,12 @@ function handleUnenvAliasedPackages(
 	);
 
 	build.onResolve({ filter: UNENV_ALIAS_RE }, (args) => {
+		const unresolvedAlias = alias[args.path];
 		// Convert `require()` calls for NPM packages to a virtual ES Module that can be imported avoiding the require calls.
-		// Note: Only applies to npm aliases as node package are handled in `handleRequireCallsToNodeJSBuiltins`
+		// Note: Does not apply to Node.js packages that are handled in `handleRequireCallsToNodeJSBuiltins`
 		if (
 			args.kind === "require-call" &&
-			alias[args.path].startsWith("unenv/runtime/npm")
+			(unresolvedAlias.startsWith("unenv/runtime/npm/") || unresolvedAlias.startsWith("unenv/runtime/mock/"))
 		) {
 			return {
 				path: args.path,
@@ -118,7 +119,7 @@ function handleUnenvAliasedPackages(
 		// Resolve the alias to its absolute path and potentially mark it as external
 		return {
 			path: aliasAbsolute[args.path],
-			external: external.includes(alias[args.path]),
+			external: external.includes(unresolvedAlias),
 		};
 	});
 

